### DISCUSSION TITLE
TypeError: float() argument must be a string or a number, not 'BaseVar'

### DISCUSSION THREAD

**Discussion by gsal** [reflex-dev/reflex#554] [Q&A]
The code below yields the error in the title.
Can somebody please help? explain?

I am sure you are going to question the structure of the program and wonder why I am not taking advantage of Pynecone intrinsic setters of state variables, etc. ....well, there is a reason, but I don't think that matters at this moment...if you think if does and need to know, I can explain; otherwise...

...if seems to me the code below is rather harmless and I don't understand why Pynecone thinks the lambda v function argument is a BaseVar and not just some string coming out of the number_input component. 

```
from pcconfig import config
import pynecone as pc

class State(pc.State):
    density: float = 0.0

def is_float(s) -> bool:
    try:
        float(s)
        return True
    except ValueError:
        return False

def update_Air(sval):
    if is_float(sval):
        State.density = float(sval)

def index():
    return pc.center(
        pc.hstack(
            pc.text("Density_at_inlet", font_size="0.8em"),
            pc.number_input(on_change=lambda v: update_Air(v), input_mode="decimal", variant="outline", font_size="0.8em"),
        )
    )

# Add state and page to the app.
app = pc.App(state=State)
app.add_page(index)
app.compile()

```

**Comment by Lendemor** [comment]
All event handlers need to be declared inside a State (or substate) so they can interact with the Vars inside this same state.

In your case, you should have :
```
class State:
    density: float = 0.0

    def update_air(self, sval):
        self.density = float(sval)
```

Also, If I'm not wrong, on change already provide the changed value  `v` so you don't need the lambda at all:
```
pc.number_input(on_change=State.update_air, ...)
```

TLDR : any method that modify a value in a State should belong to said State.

**Comment by gsal** [comment]
Thanks for that....I am learning and trying to wrap my head around this framework paradigm and the rules of the game. 

Yeah, you may be right about the changed value v, but I am about to pass additional arguments to the update_air function.
